File: 在线仓库/modelzoo-master/samples/samples_GPL/common/infer/preprocess/yolov8s_seg_preprocess.cpp
/*
 * Copyright (c) ModelZoo. 2025-2025. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "yolov8s_seg_preprocess.h"
#include "dev_interface_adapter.h"
#include "log.h"
#include "utils.h"
#include <algorithm>
#include <fstream>
#include <nlohmann/json.hpp>
#include <opencv2/opencv.hpp>
#include <string>
#include <sys/stat.h>
#include <sys/types.h>
#include <vector>

using json = nlohmann::json;
constexpr int BYTE_BIT_NUM = 8;
constexpr int FP16_BIT_NUM = 16;
constexpr int FP32_BIT_NUM = 32;
namespace Infer {
namespace Yolov8sSegNS {
    // 新增像素值归一化系数常量，替代魔鬼数字255.0
    constexpr float PIXEL_NORMALIZE_FACTOR = 255.0f;
    constexpr int CHANNEL_COUNT = 3;

    static Result ReadImgFileToBuf(cv::Mat& chwImg, TensorDesc& desc, TensorBuf& inBuf)
    {
        LOG(INFO) << "ReadImgFileToBuf: desc.dimCount " << desc.dimCount;
        const float* srcData = chwImg.ptr<float>();
        // for dlite
        if (desc.defaultStride == 0) {
            if (desc.typeSize == FP32_BIT_NUM) {
                memcpy(static_cast<float*>(inBuf.GetRawPtr()), srcData, desc.defaultSize);
                return SUCCESS;
            } else if (desc.typeSize == FP16_BIT_NUM) {
                // 计算数据量
                size_t dataCount = chwImg.total();

                // 转换为FP16
                std::vector<uint16_t> fp16Data(dataCount);
                for (size_t i = 0; i < dataCount; ++i) {
                    fp16Data[i] = FloatToHalf(srcData[i]);
                }

                memcpy(static_cast<void*>(inBuf.GetRawPtr()), fp16Data.data(), desc.defaultSize);
                return SUCCESS;
            } else {
                LOG(ERROR) << "dlite core unsupported desc.typeSize: " << desc.typeSize;
                return FAILED;
            }
        }

        // dpico核，需要处理字节对齐
        int64_t loopTimes = 1;
        for (size_t loop = 0; loop < desc.dimCount - 1; loop++) {
            loopTimes *= desc.dims[loop];
        }

        int64_t width = desc.dims[desc.dimCount - 1]; /* dims last dim is width */
        size_t dataSize = desc.typeSize / BYTE_BIT_NUM;
        size_t strideElemNum = inBuf.stride / dataSize;

        for (int64_t loop = 0; loop < loopTimes; loop++) {
            float* destPtr = static_cast<float*>(inBuf.GetRawPtr()) + loop * strideElemNum;
            const float* srcPtr = srcData + loop * width;

            if (destPtr == nullptr || srcPtr == nullptr) {
                LOG(ERROR) << "Null pointer detected";
                return FAILED;
            }

            memcpy(destPtr, srcPtr, width * dataSize);
        }
        return SUCCESS;
    }

    // Letterbox函数实现
    static cv::Mat Letterbox(const cv::Mat& img, const cv::Size& targetSize, bool scaleup)
    {
        cv::Size shape = img.size();
        const float minValue = 0.1f;

        double r = std::min(static_cast<double>(targetSize.width) / shape.width,
            static_cast<double>(targetSize.height) / shape.height);

        LOG(INFO) << "read r: " << r;
        cv::Size unpad(static_cast<int>(std::round(shape.width * r)),
            static_cast<int>(std::round(shape.height * r)));

        LOG(INFO) << "unpad r: " << unpad.width << "  " << unpad.height;
        double w = (targetSize.width - unpad.width) / 2.0f;
        double h = (targetSize.height - unpad.height) / 2.0f;
        LOG(INFO) << "w r: " << w << "  " << h;

        cv::Mat resized;
        if (shape != unpad) {
            cv::resize(img, resized, unpad, 0, 0, cv::INTER_LINEAR);
        } else {
            resized = img.clone();
        }

        int top = static_cast<int>(std::round(h - minValue));
        int bottom = static_cast<int>(std::round(h + minValue));
        int left = static_cast<int>(std::round(w - minValue));
        int right = static_cast<int>(std::round(w + minValue));
        const int padValue = 114;

        cv::Mat padded;
        cv::copyMakeBorder(resized, padded, top, bottom, left, right,
            cv::BORDER_CONSTANT, cv::Scalar(padValue, padValue, padValue));

        return padded;
    }

    static cv::Mat HwcToChw(const cv::Mat& hwcImg)
    {
        CV_Assert(!hwcImg.empty() && hwcImg.channels() == CHANNEL_COUNT);
        std::vector<cv::Mat> channels;
        cv::split(hwcImg, channels);

        cv::Mat chwMat(CHANNEL_COUNT, hwcImg.rows * hwcImg.cols, hwcImg.depth());
        chwMat = cv::Scalar(0);

        for (int channelIdx = 0; channelIdx < CHANNEL_COUNT; ++channelIdx) {
            cv::Mat flatChannel = channels[channelIdx].reshape(1, 1).clone();
            flatChannel.copyTo(chwMat.row(channelIdx));
        }

        return chwMat;
    }

    bool Yolov8sSegPreprocess(std::vector<std::string>& fileList, std::vector<TensorBuf>& tensorBufs, std::vector<TensorDesc>& tensorDescs)
    {
        // 读取配置文件
        auto cfg = ReadCfgFile("../data/cfg.txt");
        int imgHeight = std::stoi(cfg["img_height"]);
        int imgWidth = std::stoi(cfg["img_width"]);
        std::string saveBin = cfg["save_preprocess_bin"];

        // 创建保存目录
        if (!saveBin.empty() && !CreateDirectoryRecursive(saveBin)) {
            LOG(ERROR) << "Create preprocess bin dir failed";
            return false;
        }

        LOG(INFO) << "Processing " << fileList.size() << " images...";

        // 处理每个图像
        std::vector<int> imgSize = { imgWidth, imgHeight };
        for (size_t i = 0; i < fileList.size(); ++i) {
            std::string imgPath = fileList[i];
            LOG(INFO) << "imgPath: " << imgPath;
            // 提取文件名（不含路径和后缀）
            size_t start = imgPath.find_last_of("/");
            size_t end = imgPath.find_last_of(".");
            std::string outputName = imgPath.substr(start + 1, end - start - 1);
            // 读取图片文件
            cv::Mat im0 = cv::imread(imgPath);

            if (im0.empty()) {
                LOG(ERROR) << "Read image failed: " << imgPath;
                return false;
            }

            // 应用letterbox
            cv::Mat processed = Letterbox(im0, cv::Size(imgSize[0], imgSize[1]), false);

            // BGR到RGB
            cv::cvtColor(processed, processed, cv::COLOR_BGR2RGB);

            cv::Mat chwImg = HwcToChw(processed);

            if (!chwImg.isContinuous()) {
                chwImg = chwImg.clone();
            }

            // 转换为浮点数并归一化
            cv::Mat doubleTemp;
            chwImg.convertTo(doubleTemp, CV_64FC3);
            doubleTemp /= PIXEL_NORMALIZE_FACTOR;
            doubleTemp.convertTo(chwImg, CV_32FC3);

            ReadImgFileToBuf(chwImg, tensorDescs[i], tensorBufs[i]);
            LOG(INFO) << "read end: " << imgPath;

            // 保存预处理结果
            if (!saveBin.empty()) {
                std::string binPath = saveBin + "/" + outputName + ".bin";
                std::ofstream binFile(binPath, std::ios::binary);
                if (binFile.is_open()) {
                    const int floatSize = sizeof(float);
                    size_t dataSize = imgWidth * imgHeight * CHANNEL_COUNT * floatSize;
                    binFile.write((char*)chwImg.data, dataSize);
                    binFile.close();
                    // 设置文件权限
                    const int filePermission = 0755;
                    if (chmod(binPath.c_str(), filePermission) != 0) {
                        LOG(WARNING) << "Failed to set permissions for bin file: " << binPath;
                    }
                    LOG(INFO) << "Saved preprocess bin: " << binPath;
                } else {
                    LOG(WARNING) << "Failed to open bin file: " << binPath;
                }
            }
        }

        LOG(INFO) << "PreProcessing completed successfully!";
        return true;
    }
}
}

File: 在线仓库/modelzoo-master/samples/samples_GPL/built-in/yolov8s-seg/script/yolov8s_seg_preprocess.py
# Copyright (c) ModelZoo. 2025-2025. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import cv2
import numpy as np
import argparse
import json
from tqdm import tqdm


def letterbox(img, target_size=(640, 640), color=(114, 114, 114)):
    """保持比例缩放图片并填充至目标尺寸（与Ultralytics实现对齐）"""
    shape = img.shape[:2]  # 原始尺寸 (h, w)
    if isinstance(target_size, int):
        target_size = (target_size, target_size)

    # 计算缩放比例
    r = min(target_size[0] / shape[0], target_size[1] / shape[1])
    # 中心对齐填充
    new_unpad = int(round(shape[1] * r)), int(round(shape[0] * r))  # (new_w, new_h)
    dw, dh = (target_size[1] - new_unpad[0]) / 2, (target_size[0] - new_unpad[1]) / 2  # 填充量

    if shape[::-1] != new_unpad:
        img = cv2.resize(img, new_unpad, interpolation=cv2.INTER_LINEAR)
    top, bottom = int(round(dh - 0.1)), int(round(dh + 0.1))
    left, right = int(round(dw - 0.1)), int(round(dw + 0.1))

    # 填充边界
    img = cv2.copyMakeBorder(img, top, bottom, left, right,
                             cv2.BORDER_CONSTANT, value=color)
    # 格式转换: BGR->RGB, 归一化, 转置为CHW格式
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = img.astype(np.float32) / 255.0
    img = img.transpose(2, 0, 1)
    img = np.ascontiguousarray(img)
    img = np.expand_dims(img, axis=0)  # 添加batch维度
    return img


def image_to_bin(image_path, output_bin_path, target_size=(640, 640)):
    """将图片转换为bin文件（符合YOLOv8输入要求）"""
    image = cv2.imread(image_path)
    if image is None:
        print(f"无法读取图片 {image_path}")
        return False

    # 预处理
    image = letterbox(image, target_size)

    # 保存为bin文件
    try:
        image.tofile(output_bin_path)
        return True
    except Exception as e:
        print(f"保存bin文件失败 {output_bin_path}: {e}")
        return False


def process_images(file_list_path, output_dir, bin_list_file, target_size=(640, 640), max_count=5000):
    """处理图片并生成bin文件列表"""
    os.makedirs(output_dir, exist_ok=True)

    if not os.path.exists(file_list_path):
        print(f"错误: 文件列表 {file_list_path} 不存在")
        return

    with open(file_list_path, 'r') as f:
        try:
            data = json.load(f)
            image_files = [item[0] for item in data.get('fileList', [])]
        except json.JSONDecodeError as e:
            print(f"解析JSON文件失败: {e}")
            return

    process_files = image_files[:max_count]

    with open(bin_list_file, 'w') as f_list:
        for filename in tqdm(process_files, desc="处理进度"):
            if not os.path.exists(filename):
                print(f"文件不存在，跳过: {filename}")
                continue

            base_name = os.path.splitext(os.path.basename(filename))[0]
            output_path = os.path.join(output_dir, f"{base_name}.bin")

            if image_to_bin(filename, output_path, target_size):
                abs_output_path = os.path.abspath(output_path)
                f_list.write(f"{abs_output_path}\n")


def main():
    parser = argparse.ArgumentParser(description='YOLOv8s-seg图像预处理脚本')
    parser.add_argument('--file_list', default='../data/file_list.json',
                        help='包含图片路径的JSON文件路径')
    parser.add_argument('--output_dir', default='../data/preprocess/bin',
                        help='输出bin文件路径')
    parser.add_argument('--target_size', type=int, nargs=2, default=[640, 640],
                        help='目标尺寸，格式为 宽 高')
    parser.add_argument('--max_count', type=int, default=5000,
                        help='最大处理图片数量')

    args = parser.parse_args()
    process_images(
        file_list_path=args.file_list,
        output_dir=args.output_dir,
        bin_list_file=os.path.dirname(args.output_dir) + '/file_list.txt',
        target_size=tuple(args.target_size),
        max_count=args.max_count
    )
    print(f"预处理完成，bin文件已保存至 {args.output_dir}")


if __name__ == "__main__":
    main()
